px4_ros2_cpp/include/px4_ros2/utils/geodesic.hpp

```diff
@@ -0,0 +1,250 @@
+/****************************************************************************
+ * Copyright (c) 2024 PX4 Development Team.
+ * SPDX-License-Identifier: BSD-3-Clause
+ ****************************************************************************/
+
+#pragma once
+
+#include <cmath>
+#include <Eigen/Eigen>
+#include <rclcpp/rclcpp.hpp>
+#include <px4_msgs/msg/vehicle_local_position.hpp>
+#include <px4_ros2/common/context.hpp>
+
+namespace px4_ros2
+{
+
+class MapProjectionImpl;
+
+class MapProjection
+{
+public:
+  explicit MapProjection(Context & context);
+
+  ~MapProjection() = default;
+
+  /**
+   * @return true, if the map reference has been initialized before
+   */
+  bool isInitialized() const;
+
+  /**
+   * Transform a point in the geographic coordinate system to the local
+   * azimuthal equidistant plane using the projection
+   *
+   * @param global_position lat [deg], lon [deg] (degrees: 8.1234567°, not 81234567°)
+   * @return the point in local coordinates as north, east [m]
+   */
+  Eigen::Vector2f globalToLocal(const Eigen::Vector2d & global_position) const;
+
+  /**
+   * Transform a point in the geographic coordinate system to the local
+   * azimuthal equidistant plane using the projection
+   *
+   * @param global_position lat [deg], lon [deg], alt AMSL [m] (degrees: 8.1234567°, not 81234567°)
+   * @return the point in local coordinates as north, east, down [m]
+   */
+  Eigen::Vector3f globalToLocal(const Eigen::Vector3d & global_position) const;
+
+  /**
+   * Transform a point in the local azimuthal equidistant plane to the
+   * geographic coordinate system using the projection
+   *
+   * @param local_position north, east [m]
+   * @return the point in geographic coordinates as lat [deg], lon [deg] (degrees: 8.1234567°, not 81234567°)
+   */
+  Eigen::Vector2d localToGlobal(const Eigen::Vector2f & local_position) const;
+
+  /**
+   * Transform a point in the local azimuthal equidistant plane to the
+   * geographic coordinate system using the projection
+   *
+   * @param local_position north, east, down [m]
+   * @return the point in geographic coordinates as lat [deg], lon [deg], alt AMSL [m] (degrees: 8.1234567°, not 81234567°)
+   */
+  Eigen::Vector3d localToGlobal(const Eigen::Vector3f & local_position) const;
+
+private:
+  /**
+   * @throw runtime error if class instance is not initalized
+   */
+  void assertInitalized() const;
+
+  rclcpp::Node & _node;
+  std::shared_ptr<MapProjectionImpl> _map_projection_math;
```

Comment:
I initially struggled to do this until I found this thread: https://stackoverflow.com/questions/13414652/forward-declaration-with-unique-ptr. I'm not sure I fully get it and you may have already known, thought I'd share just in case

Reply:
Ah, no, I was not aware of that.